### DISCUSSION TITLE
Tracing/Grafana check

### DISCUSSION THREAD

**Discussion by dimitarshenkov** [kiali/kiali#5190] [General]
Hi , I have internal Jaegr installed  and configuration like 
![image](https://user-images.githubusercontent.com/99493924/173292365-fb1c3c4e-1c8c-451a-ad14-ef93a65edcaa.png)
I have some dashboards installed too.
How can I check that in cluster Jaeger from Kiali UI?
I can see only external link right now
![image](https://user-images.githubusercontent.com/99493924/173292637-848af3dd-f07c-46ef-bd42-ca761796775f.png)

**Comment by lucasponce** [comment]
> How can I check that in cluster Jaeger from Kiali UI?

If you have enabled the "traces" tab in any of the locations of the Kiali UI, then you have established a connection.

If you had any problems, you should see some warnings in the Kiali "masthead" with more details.

**Comment by dimitarshenkov** [comment]
Thanks , and what about  in_cluster_url: http://grafana.istio-system:3000    , where can I see it
I have dashboards in Grafana not in Jaeger , sorry for the mistake

**Reply by lucasponce** [reply]
It's the same, if not populated, or it can't connect it should show an info/warning message in the masthead.

**Reply by dimitarshenkov** [reply]
I mean how can I see its in use in UI, which tab ?

**Reply by lucasponce** [reply]
Ah, you have several places:
- From the "Overview" page, if you see the control-plane namespace (i.e. istio-system), you should see the links to the predefined Grafana dashboard for Istio
- From the "Details" pages, any of the "Inbound/Outbound Metrics" should have a Grafana link to point to the workload/services dashboard.

Note that the Grafana URLs are dynamic, and kiali has access to the Grafana service to ask for the data needed to build those links that should be accessible from the browser.

**Comment by dimitarshenkov** [comment]
I have link to external garafana but where is used internal/in cluster ?
![image](https://user-images.githubusercontent.com/99493924/173353236-211a8fd1-0e1d-4263-8b42-a1aecdd582b5.png)

**Reply by lucasponce** [reply]
https://kiali.io/docs/configuration/kialis.kiali.io/#.spec.external_services.grafana.in_cluster_url
(for internal queries between Kiali pod -> Grafana service)

https://kiali.io/docs/configuration/kialis.kiali.io/#.spec.external_services.grafana.url
(for external queries, from the browser to the public URL where Grafana is running)

**Reply by lucasponce** [reply]
if no set, Kiali assumes that the defaults are a similar installation as offered in the Istio addons.

**Comment by dimitarshenkov** [comment]
I know how to configure internal , just cant understand where it is in use in Kiali , where can I see internal dashboards which are scraped by internal Grafana

**Reply by jmazzitelli** [reply]
Take a look at the screen shot here - see the grafana link? That's one place - is this what you are asking? 

https://user-images.githubusercontent.com/9693382/95454265-822c6e80-096c-11eb-9c7c-6260f2e67cc7.png

**Reply by lucasponce** [reply]
In the Overview page:
![image](https://user-images.githubusercontent.com/1662329/173360784-f0e9b7ed-1b38-4e28-9ad8-1c7defe3b219.png)

In the Service/Workload details page:
![image](https://user-images.githubusercontent.com/1662329/173360924-37010a8d-9ba1-4798-b735-810be25aadc4.png)

Note the "dynamic" URL of Grafana on workload/service details, that's what Kiali needs to query before it can render the url.

**Reply by dimitarshenkov** [reply]
@lucasponce  this links sending me to external garfana
![image](https://user-images.githubusercontent.com/99493924/173377056-3f15750f-9715-4622-9fb0-f47dc367625c.png)

**Reply by dimitarshenkov** [reply]
@lucasponce So without internal Garfana I wont have graphics in  the Service/Workload details page ?

**Reply by dimitarshenkov** [reply]
> Take a look at the screen shot here - see the grafana link? That's one place - is this what you are asking?
> 
> https://user-images.githubusercontent.com/9693382/95454265-822c6e80-096c-11eb-9c7c-6260f2e67cc7.png

No Im not asking that :)

**Reply by lucasponce** [reply]
> So without internal Garfana I wont have graphics in the Service/Workload details page ?

That's correct. The URLs for workloads and services are dynamic, so it needs an internal call to build it

**Comment by jmazzitelli** [comment]
And here's the code that uses it - it needs this in cluster URL to ask Grafana for some information about the dashboards:

* https://github.com/kiali/kiali/blob/5bcdd2f372ff2fcca4746e321bd9e93e677e1b0c/status/discover.go#L55
* https://github.com/kiali/kiali/blob/cbabc4cf1cde59e471577147aeec4823f4b39ca4/business/grafana.go#L115-L197

**Reply by dimitarshenkov** [reply]
I still have metrics even  without internal Grafana addon installed
![image](https://user-images.githubusercontent.com/99493924/173381150-a5276d17-d6de-4536-bee1-16d657623639.png)

**Comment by jshaughn** [comment]
Grafana dashboards are not embedded in Kiali, they will always be an external link.  Kiali has its own charts in the detail tabs, as you posted above.

**Reply by dimitarshenkov** [reply]
My simple question is , why we need this config in_cluster_url: http://grafana.istio-system:3000 ?

**Reply by lucasponce** [reply]
I've answered below 
![image](https://user-images.githubusercontent.com/1662329/173589970-6b648f36-284a-46ea-98bf-92a0d05ff8a1.png)
that url to link a workload with a grafana dashboard is not static, it's dynamic and Kiali needs to ask Grafana "what is the URL for the workload A or B?" otherwise that link would not work correctly

**Reply by lucasponce** [reply]
That "question" to Grafana to build the "dynamic" URL happens internally between Kiali pod and Grafana service

**Reply by dimitarshenkov** [reply]
@lucasponce I dont see any difference in Kiali if  incluster grafana is installed or not, thats the problem

**Reply by lucasponce** [reply]
if that Grafana internal URL is not set, kiali has some basic "auto-discovery" logic to try to guess the internal URL (i.e. grafana.istio-system), that logic was set up as some users expose the grafana service through other services.

I.e. if you put that in-cluster URL with some incorrect URL, it should fail.

**Reply by jmazzitelli** [reply]
> My simple question is , why we need this config in_cluster_url

I linked to the code ^^ that answers your question. See https://github.com/kiali/kiali/discussions/5190#discussioncomment-2938428

**Reply by dimitarshenkov** [reply]
@lucasponce 
 dont see any difference in Kiali --> GUI   if incluster grafana is installed or not, thats the problem, no new stats or images

**Reply by lucasponce** [reply]
Yes, I think there is an edge in the "discover" logic that uses the public URL when this is defined:
![image](https://user-images.githubusercontent.com/1662329/173759524-117abb56-ec3b-4b86-ba4f-dfe632504ee9.png)

**Reply by dimitarshenkov** [reply]
Ohhh thats the case then thanx, will test

**Reply by lucasponce** [reply]
Thanks :). Autodiscovery logic is always tricky. This was set to make it easier for users and no need to explicitly configure extra steps, but in some corner cases, it may introduce some noise.

It's traditionally pretty stable, but let us know if something is not working.
